[Sema] Improve Optional-to-Any diagnostics for collections, nested optionals & IUOs

hamishknight opened the pull request:
- When implicitly coercing a collection of optional elements to a collection of less-optional `Any` elements, e.g `[Any?]` to `[Any]`, emit a slightly better diagnostic:

  > Expression implicitly coerced from '[Any?]' to '[Any]'

  and allow for silencing of the diagnostic with an explicit coercion, e.g `as [Any]`.

- When implicitly coercing a nested optional to a less optional `Any`, e.g `Any??` to `Any?`, emit a slightly better diagnostic:

  > Expression implicitly coerced from 'Any??' to 'Any?'

  and allow for silencing of the diagnostic with an explicit coercion, e.g `as Any?`.

(both diagnostics were previously "Expression implicitly coerced from 'Any?' to Any", and neither could be silenced by an explicit coercion to their destination types)

In addition, this PR enables Optional-to-Any warnings for IUOs, e.g the implicit coercion from `Int!` to `Any` now produces a warning. Previously a warning wasn't emitted despite the optional being implicitly erased to `Any` (but warnings were emitted in some other cases, such as `Int?!` to `Any?`; this PR fixes that inconsistency).

Resolves [SR-2928](https://bugs.swift.org/browse/SR-2928) & [SR-6859](https://bugs.swift.org/browse/SR-6859).

hamishknight committed:
[Sema] Refactor Optional-to-Any diagnostic logic (NFC)

hamishknight committed:
[Sema] Change Optional-to-Any diagnostics to take a variable destination type

This will let us customise the diagnostic for collections and nested optionals.

hamishknight committed:
[Sema] Improve Optional-to-Any diagnostics for nested optionals

When implicitly coercing a nested optional to a less optional `Any`, e.g `Any??` to `Any?`, emit a slightly better diagnostic:

Expression implicitly coerced from 'Any??' to 'Any?'

and allow for silencing of the diagnostic with an explicit coercion, e.g `as Any?`.

hamishknight committed:
[Sema] Improve Optional-to-Any diagnostics for collections

When implicitly coercing a collection of optional elements to a collection of less-optional `Any` elements, e.g `[Any?]` to `[Any]`, emit a slightly better diagnostic:

Expression implicitly coerced from '[Any?]' to '[Any]'

and allow for silencing of the diagnostic with an explicit coercion, e.g `as [Any]`.

hamishknight commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -3449,7 +3449,222 @@ static void diagnoseUnintendedOptionalBehavior(TypeChecker &TC, const Expr *E,
+      // to Any coercions (this doesn't take into consideration implicit
+      // coercions such as Any?! to Any?, however; we warn on those).
+      if (subExpr->getType()->getImplicitlyUnwrappedOptionalObjectType())
+        return;
```
It would be great if we could remove this, which would enable warnings for implicit IUO to Any coercions, such as shown in the `dontWarnIUOToAnyCoercion` function in the tests (which should really produce warnings). With IUOs being removed from the type system, I believe this should happen at *some* point; but not sure if it's okay to just enable them this PR.

rudkx replied:
I would say feel free to warn about the IUOs as well. It will fall out from my last commit to always generate Optional<T> from '!', but there is no reason not to do so now, and it's really inconsistent that we haven't been emitting the warnings in that case as well.

hamishknight replied:
@rudkx Thanks, I'll go ahead and make that change then. It looks like we have 3 tests that make such implicit IUO coercions, two of which look like they can be updated with `as Any` without affecting the test ([1](https://github.com/apple/swift/blob/d23e2bce6274d3b61c9c10a17f97a27da87d7fdf/test/Constraints/casts_objc.swift#L37) & [2](https://github.com/apple/swift/blob/d23e2bce6274d3b61c9c10a17f97a27da87d7fdf/test/Constraints/existential_metatypes.swift#L90)). Though I'm not sure about [the third one](https://github.com/apple/swift/blob/d23e2bce6274d3b61c9c10a17f97a27da87d7fdf/test/Constraints/diag_ambiguities.swift#L52), can it be updated with an explicit `as Any` or `as [Any]`, or would it be best to just add an expected warning?

rudkx replied:
I would add the expected warning on the last one, but yes I think you're probably right that adding the coercions for the others is fine.

hamishknight replied:
@rudkx Thanks, I've gone ahead and done that now.

hamishknight committed:
[Sema] Enable Optional-to-Any diagnostics for IUOs

For example, the implicit coercion from `Int!` to `Any` now produces a warning. Previously a warning wasn't emitted despite the optional being implicitly erased to `Any` (but warnings were emitted in some other cases, such as `Int?!` to `Any?`; this commit fixes that inconsistency).

rudkx commented:
@swift-ci Please smoke test

hamishknight committed:
[Sema] Update Optional-to-Any tests for #14299

As IUOs types are no longer generated, we don't show them in diagnostics.

hamishknight commented:
@rudkx I've updated the tests for #14299 (such that the diagnostics only show strong optional, not IUO types), could you please kick off the CI bot again?

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
I'll take a closer look at the changes as well.

rudkx commented:
LGTM. I'll merge. That will result in conflicts in another of my PR but I'll get that fixed up ASAP.
